online-shop/src/reducers/shopping.tsx

```diff
@@ -0,0 +1,204 @@
+      return {
+        cartProducts: insertInCart(loadAction.product, [...state.cartProducts]),
+        hasProducts: true,
+        responseFromBackend: 0,
+        showModal: false,
+        orderItem: state.orderItem,
+        modalText: state.modalText,
+        modalTitle: state.modalTitle,
+        json:state.json
+      };
+    case "INCREASE-QUANTITY":
+      const addAction: QuantityUP = action as QuantityUP;
+      return {
+        cartProducts: increaseQuantity(addAction.product, state.cartProducts),
+        hasProducts: true,
+        responseFromBackend: 0,
+        showModal: false,
+        orderItem: state.orderItem,
+        modalText: state.modalText,
+        modalTitle: state.modalTitle,
+        json:state.json
+      };
+    case "DECREASE-QUANTITY":
+      const eraseAction: QuantityDown = action as QuantityDown;
+      return {
+        cartProducts: decreaseQuantity(eraseAction.product, state.cartProducts),
+        hasProducts: true,
+        responseFromBackend: 0,
+        showModal: false,
+        orderItem: state.orderItem,
+        modalText: state.modalText,
+        modalTitle: state.modalTitle,
+        json:state.json
+      };
+    case "DELETE-PRODUCT-FROM-CART":
+      const deleteAction: EraseItem = action as EraseItem;
+      return {
+        cartProducts: deleteItemFromCart(deleteAction.product, [
+          ...state.cartProducts
+        ]),
+        hasProducts: true,
+        responseFromBackend: 0,
+        showModal: false,
+        orderItem: state.orderItem,
+        modalText: state.modalText,
+        modalTitle: state.modalTitle,
+        json:state.json
+      };
+
+    case "CHECK-OUT":
+      const checkOutAction: CheckOut = action as CheckOut;
+      return {
+        cartProducts: clearCart([...checkOutAction.cartProducts]),
+        hasProducts: true,
+        responseFromBackend: state.responseFromBackend,
+        showModal: true,
+        orderItem: state.orderItem,
+        modalText: checkOutAction.modalText,
+        modalTitle: checkOutAction.modalTitle,
+        json:state.json
+      };
+      case "FETCH-ORDER":
+          const fetchAction: FetchOrder = action as FetchOrder;
+          return {
+            ...state,
+            cartProducts:fetchAction.cartProducts,
+            modalText:fetchAction.modalText,
+            modalTitle:fetchAction.modalTitle,
+            json:fetchAction.json
+          };
+
+    default:
+      return state;
+  }
+}
+
+let insertInCart = function addProductsToCart(
+  newCartProduct: IProduct,
+  crtCartItems: ICartProduct[]
+): ICartProduct[] {
+  let contor: number = 0;
+  let i: number;
+  let productInCart: boolean = false;
+  let cartItemToInsert: ICartProduct = {} as any;
+  let newCartItemsArray: ICartProduct[] = [];
+
+  for (i = 0; i < crtCartItems.length; i++) {
+    if (crtCartItems[i].product.id === newCartProduct.id) {
+      contor = crtCartItems[i].quantity;
+      contor++;
+      crtCartItems[i].quantity = contor;
```

Comment:
You should avoid modifying the old state, instead you should create a new state object. It would be a lot easier if you stored the products in a map / object (and not an array).